File: LeetCode/Python Solutions/Number of Longest Increasing Subsequence/number.py
from  typing import List


class Solution:
    def findNumberOfLIS(self, nums: List[int]) -> int:
        n = len(nums)
        if n == 0:
            return 0
        
        lengths = [0] * n  # lengths[i] = longest ending in nums[i]
        counts = [1] * n  # count[i] = number of longest ending in nums[i]
        
        for i, num in enumerate(nums):
            for j in range(i):
                if nums[j] < num:
                    if lengths[j] >= lengths[i]:
                        lengths[i] = 1 + lengths[j]
                        counts[i] = counts[j]
                    elif lengths[j] + 1 == lengths[i]:
                        counts[i] += counts[j]
        
        longest = max(lengths)
        return sum(c for l, c in zip(lengths, counts) if l == longest)

    
# Tests:
if __name__ == '__main__':
    assert Solution().findNumberOfLIS([1,3,5,4,7]) == 2
    assert Solution().findNumberOfLIS([2,2,2,2,2]) == 5
    assert Solution().findNumberOfLIS([1,2,4,3,5,4,7,2]) == 3
    print('All Passed!')


# Submission Result:
# 223/223 cases passed (180 ms)
# Your runtime beats 99.09 % of python3 submissions
# Your memory usage beats 100 % of python3 submissions (12.8 MB)ß

File: LeetCode/Java Solutions/Number of Longest Increasing Subsequence/number.java
import java.util.Arrays;

public class Solution {
    public int findNumberOfLIS(int[] nums) {
        // get the length of the array
        int n = nums.length;
        
        // if the length is zero, return 0
        if (n == 0) {
            return 0;
        }
        
        // lengths[i] = longest ending in nums[i]
        int[] lengths = new int[n];  
        Arrays.fill(lengths, 0);

        // count[i] = number of longest ending in nums[i]
        int[] counts = new int[n];   
        Arrays.fill(counts, 1);
        
        // loop through each number in the array
        for (int i = 0; i < n; i++) {
            // for each number, check the numbers before it
            for (int j = 0; j < i; j++) {
                // if the current number is larger than the previous one
                if (nums[j] < nums[i]) {
                    // if the length of the longest increasing subsequence ending with nums[j] is larger or equal to that with nums[i]
                    if (lengths[j] >= lengths[i]) {
                        lengths[i] = 1 + lengths[j]; // set the length of the longest increasing subsequence ending with nums[i] to that of nums[j] plus 1
                        counts[i] = counts[j]; // set the count of the longest increasing subsequence ending with nums[i] to that of nums[j]
                    } else if (lengths[j] + 1 == lengths[i]) { // else if the length of the longest increasing subsequence ending with nums[j] plus 1 equals to that of nums[i]
                        counts[i] += counts[j]; // add the count of the longest increasing subsequence ending with nums[j] to that of nums[i]
                    }
                }
            }
        }

        // find the maximum length
        int longest = 0;
        for (int length : lengths) {
            longest = Math.max(longest, length);
        }

        // find the number of longest increasing subsequence
        int total = 0;
        for (int i = 0; i < n; i++) {
            if (lengths[i] == longest) {
                total += counts[i];
            }
        }
        return total;
    }

    // tests
    public static void main(String[] args) {
        Solution solution = new Solution();
        assert solution.findNumberOfLIS(new int[] {1,3,5,4,7}) == 2;
        assert solution.findNumberOfLIS(new int[] {2,2,2,2,2}) == 5;
        assert solution.findNumberOfLIS(new int[] {1,2,4,3,5,4,7,2}) == 3;
        System.out.println("All Passed!");
    }
}
